core/invoker/src/main/scala/whisk/core/invoker/InvokerReactive.scala

```diff
@@ -72,31 +69,12 @@ class InvokerReactive(config: WhiskConfig, instance: InstanceId, producer: Messa
             consumer, maximumContainers, 500.milliseconds, processActivationMessage)
     })
 
-    /** Initialize container clients */
-    implicit val docker = new DockerClientWithFileAccess()(ec)
-    implicit val runc = new RuncClient(ec)
-
-    /** Cleans up all running wsk_ containers */
-    def cleanup() = {
-        val cleaning = docker.ps(Seq("name" -> s"wsk${instance.toInt}_"))(TransactionId.invokerNanny).flatMap { containers =>
-            val removals = containers.map { id =>
-                runc.resume(id)(TransactionId.invokerNanny).recoverWith {
-                    // Ignore resume failures and try to remove anyway
-                    case _ => Future.successful(())
-                }.flatMap {
-                    _ => docker.rm(id)(TransactionId.invokerNanny)
-                }
-            }
-            Future.sequence(removals)
-        }
-
-        Await.ready(cleaning, 30.seconds)
-    }
-    cleanup()
-    sys.addShutdownHook(cleanup())
 
     /** Factory used by the ContainerProxy to physically create a new container. */
-    val containerFactory = SpiLoader.get[ContainerFactoryProvider]().getContainerFactory(actorSystem, logging, config).createContainer _
+    val containerFactory = SpiLoader.get[ContainerFactoryProvider].getContainerFactory(actorSystem, logging, config, instance)
+    val containerFactoryFunction = containerFactory.createContainer _
+    containerFactory.cleanup()
+    sys.addShutdownHook(containerFactory.cleanup())
```

Comment:
I think this just effectively runs `cleanup` twice at startup. Shouldn't it be `sys.addShutdownHook { containerFactory.cleanup() }` instead?

Reply:
This is the same syntax that was there before; I tested it and it does the right thing - runs once at startup, and once at shutdown.

Reply:
For posterity:

```scala
def addShutdownHook(body: => Unit): ShutdownHookThread = ShutdownHookThread(body)
```

`body` is a by-name argument, which is only evaluated once its used. I agree it looks like the function is actually called, but braces or parentheses do no make a difference in Scala here.

Reply:
Thanks for the explanation!